### ISSUE TITLE
SchemaPrinter doesn't work consistently with OneOf versus Deprecated directives

### ISSUE THREAD

**Issue by arlampin** [graphql-java/graphql-java#3357]
**Describe the bug**
Using
```
SchemaPrinter.ExcludeGraphQLSpecifiedDirectivesPredicate
``` 
works differently regarding OneOf and Deprecated directives. With `ExcludeGraphQLSpecifiedDirectivesPredicate`, both directive definitions are removed from generated file, but with `OneOf`, also directive's usage is removed from types.

So for example
```
input SomeType @oneOf {
  field1: String
  field2: Int
}
```
becomes
```
input SomeType {
  field1: String
  field2: Int
}
```

This doesn't happen with Deprecated, they are still present in the schema.

Also I didn't test with other directives.

**To Reproduce**
OneOf Directive is added with:
```
newInputObject()
            .withDirective(Directives.OneOfDirective)
            ...
```
and deprecated with:
```
newFieldDefinition()
            .deprecate("This field is deprecated")
            ...
```
`SchemaPrinter` is used as:
```
new SchemaPrinter(SchemaPrinter.Options.defaultOptions()
            .includeDirectives(SchemaPrinter.ExcludeGraphQLSpecifiedDirectivesPredicate))
```

**Comment by dondonz** [comment]
Hello, thanks for checking out `@oneOf` so soon after release!

This is intended behaviour, the `@deprecated` directive is treated differently, it's always printed.

See the `SchemaPrinter` class, these snippets are from the latest commit on master 
https://github.com/graphql-java/graphql-java/blob/ea4414f8133a163aec5ee2b3fd1318346019afc7/src/main/java/graphql/schema/idl/SchemaPrinter.java#L878
and
https://github.com/graphql-java/graphql-java/blob/ea4414f8133a163aec5ee2b3fd1318346019afc7/src/main/java/graphql/schema/idl/SchemaPrinter.java#L912

**Comment by dondonz** [comment]
@arlampin seeing as this schema printer code was written a while ago, open to ideas. Would you prefer to see `@deprecated` always included, or not included in the schema printer?

**Comment by arlampin** [comment]
I think all three build-in schema directives (`@deprecated`, `@specifiedBy` and now `@oneOf`) should work the same way: if excluded through `ExcludeGraphQLSpecifiedDirectivesPredicate` their definitions should be excluded, but not usage.

This would be consistent with graphql spec:

> When representing a GraphQL schema using the type system definition language any built-in directive may be omitted for brevity.

Sure, `@oneOf` is new and not yet in the spec, and it might make sense to exclude it from the list of spec defined directives for the time being. But that wouldn't change what I wrote above with regards to `@deprecated`/`@specifiedBy`.

**Comment by bbakerman** [comment]
On reflection I think @arlampin is right that ExcludeGraphQLSpecifiedDirectivesPredicate should work for ANY directive including @deprecated.  

My line of reasoning was that @deprecated is special (and it is in that it drives a very specific behavior in Introspection) but a true / false predicate is more correct if it handles all cases

**Comment by arlampin** [comment]
Just to be clear, I would suggest that `ExcludeGraphQLSpecifiedDirectivesPredicate` works identically on all spec defined directives. But one problem with `SchemaPrinter` is that its `.includeDirectives(Predicate<String>)` affects both directive definitions and directive usage. I think ideally it should be possible to prevent `SchemaPrinter` from printing spec defined directive definitions, as that's allowed by the spec itself.

**Comment by dondonz** [comment]
Closed by #3387, thanks for reporting!
